### PR TITLE
Ruby on rails Capstone Project

### PR DESCRIPTION
**Pull request by Ashnaali3255**
In The Budget App, the following were implemented:
### Interactions

**Splash screen**

- A simple page with the name of your app (yes, you need to choose one), and links to the sign up and log in pages.
- Sign up and log in pages
- The user should be able to register in the app with full name, email and password (all mandatory).
- The user can log into the app using email and password.
- If the user is not logged in, they can't access pages that require the user to be logged in (all the pages described below).

**Home page (categories page)**

- When the user logs in, they are presented with the categories page.
- For each category, the user can see their name, icon and the total amount of all the transactions that belongs to that category.
- When the user clicks (or taps) on a category item, the application navigates to the transactions page for that category.
- There is a button "add a new category" at the bottom that brings the user to the page to create a new category.

**Transactions page**

- For a given category, the list of transactions is presented, ordered by the most recent.
- At the top of the page the user could see the total amount for the category (sum of all of the amounts of the transactions in that category).
- There is a button "add a new transaction" at the bottom that brings the user to the page to create a new transaction.
- When the user clicks on the "Back" button (<), the user navigates to the home page.
- Add a new category page
- The user fills out a form to create a new category, indicating their name and icon (both mandatory).
- The user clicks (or taps) the "Save" button to create the new category, and is taken to the home page on success.
- When the user clicks on the "Back" button (<), the user navigates to the home page.
- "Add a new transaction" page
- The user fills out a form to create a new transaction with:
- name (mandatory)
- amount (mandatory)
- categories (mandatory at least one)
- The user click (or taps) the "Save" button to create the new transaction, and is taken to the transactions page for that category.
- When the user clicks on the "Back" button (<), the user navigates to the transactions page for that category.

**Testing requirements**
Create unit and integration tests for all the most important components of your RoR application.
**Technical requirements**

- You should use Postgres as your database.
- You should use devise for authentication.
- You should validate all user input to make sure that anyone with bad intentions cannot compromise your app.
- You can use a view template engine of your choice (.erb, .slim, .haml).
- The project should be deployed and accessible online.
- Your database schema should reflect the following structure:

### CODE REVIEW

File: app/views/devise/registrations/new.html.erb [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    <%= f.email_field :email, autofocus: true, autocomplete: "email" %>
+  </div>
+
+  <div class="field">
+    <%= f.label :password %>
+    <% if @minimum_password_length %>
+    <em>(<%= @minimum_password_length %> characters minimum)</em>
+    <% end %><br />
+    <%= f.password_field :password, autocomplete: "new-password" %>
+  </div>
```
Comment (codepantha) [inline app/views/devise/registrations/new.html.erb]: - [ ] Please ensure that the register and login pages are styled according to the given design guidelines. This is an important requirement. At the moment, it looks as though no styles were applied to the aforementioned pages 🙏 
### Design Guidelines
image


### Your Design
image

File: app/views/categories/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+<% @categories.each do |category| %>
+    <a href="/categories/<%= category.id %>", id ="category_<%= category.id %>">
+        <%= render category%>
+    </a>
+<% end %>
+</div>
+
+
+
+<%= render "categories/button_category_spendings"%>
```
Comment (codepantha) [inline app/views/categories/index.html.erb:20]: - [ ] Please let's try to style the categories index page according to the given design guidelines. Please note these pointers below 👇 
1. The general layout should be the same or closely identical
2. Please use the same font family, color, and weight
3. Please ensure that the "Create New Category" button is styled exactly as in the design guidelines
### Design Guidelines
image

### Your Design
image

File: app/views/categories/show.html.erb [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+  <div class="card-body d-flex justify-content-end">
+     <h5 class="card-title">Total Spent: $<%= @total_spent %></h5>
+  </div>
+</div>
+
+<%= render @spendings %>
+
+
+
+<%= render "categories/button_category_spendings"%>
```
Comment (codepantha) [inline app/views/categories/show.html.erb]: - [ ] Please let's try to style the categories show page (transactions list) according to the given design guidelines. Please note these pointers below 👇 
1. The general layout should be the same or closely identical
2. Please get rid of the horizontal scrolling on mobile devices as it is not great for user experience (UX) 🙏 
3. Please use the same font family, color, and weight
4. Please ensure that the "Create New Category" button is styled exactly as in the design guidelines
5. There should be a "Back" button (<) at the top left of the page (navigation) such that if the user clicks on the "Back" button (<), the user navigates to the home page
### Design Guidelines
image

### Your Design
image

File: app/views/categories/new.html.erb [hunk 4]
```diff
@@ -0,0 +1,14 @@
+<%= render "layouts/navigation"%>
```
Comment (codepantha) [inline app/views/categories/new.html.erb:1]: - [ ] Please add a Back (<) button to the top left navigation of the page on the new category page. When a user clicks this button, they should get navigated to the homepage 👍 

image

File: app/views/categories/new_spending.html.erb [hunk 5]
```diff
@@ -0,0 +1,24 @@
+<%= render "spendings/spending_nav"%>
+
```
Comment (codepantha) [inline app/views/categories/new_spending.html.erb:2]: - [ ] Please add a Back (<) button to the top left navigation of the page on the new_spending page. When a user clicks this button, they should get navigated to the transactions page for that category👍 
image

File: spec/integration/categories_spec.rb [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+  end
+end
+
+RSpec.describe 'User', type: :feature do
+  let(:user) { FactoryBot.create(:user) }
+
+  before do
+    sign_in user
+  end
+end
```
Comment (codepantha) [inline spec/integration/categories_spec.rb]: - [ ] You have not written any integration tests here that test the functionality of your categories. Please try to write robust integration tests here 🙏

File: spec/integration/spendings_spec.rb [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+  end
+end
+
+RSpec.describe 'User', type: :feature do
+  let(:user) { FactoryBot.create(:user) }
+  let(:category) { FactoryBot.create(:category, user:) }
+  before do
+    sign_in user
+  end
+end
```
Comment (codepantha) [inline spec/integration/spendings_spec.rb]: - [ ] You have not written any integration tests here that test the functionality of your spendings. Please try to write robust integration tests here 🙏

**Review by codepantha — CHANGES_REQUESTED** [review]
# STATUS: CHANGES REQUIRED ♻️ 

## Hi @Ashnaali3255 👋, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!



## Highlights 🎯
✔️ Descriptive PR
✔️ Used GitFlow correctly 📌 
✔️ Good commit history
✔️ No linter errors

## Required Changes ♻️
_Check the comments under the review._

- [ ] Please ensure that your demo video  📹 contains a highlight about a piece of code that you finds interesting or are proud of making 👍

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- n/a

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/views/devise/registrations/new.html.erb (hunk 1 above)
Comment (Ashnaali3255) [reply app/views/devise/registrations/new.html.erb]: fixed:
image

Comment (Ashnaali3255) [reply app/views/devise/registrations/new.html.erb]: Login:
image

File: app/views/categories/index.html.erb (hunk 2 above)
Comment (Ashnaali3255) [reply app/views/categories/index.html.erb:20]: fixed:
image

File: app/views/categories/show.html.erb (hunk 3 above)
Comment (Ashnaali3255) [reply app/views/categories/show.html.erb]: Fixed:
image

File: app/views/categories/new.html.erb (hunk 4 above)
Comment (Ashnaali3255) [reply app/views/categories/new.html.erb:1]: fixed:
image

File: app/views/categories/new_spending.html.erb (hunk 5 above)
Comment (Ashnaali3255) [reply app/views/categories/new_spending.html.erb:2]: Fixed:
image

File: spec/integration/spendings_spec.rb (hunk 7 above)
Comment (Ashnaali3255) [reply spec/integration/spendings_spec.rb]: fixed:
image

File: spec/integration/categories_spec.rb (hunk 6 above)
Comment (Ashnaali3255) [reply spec/integration/categories_spec.rb]: fixed:
image

**Review by franclobo — APPROVED** [review]
Hi @Ashnaali3255, 

# STATUS: APPROVED 😃 

leonardo-dicaprio-leo-dicaprio

In my opinion, your project is ready for the presentation!  There is nothing else to say other than... it's time to merge it :shipit: 

## Highlights

- No inter errors ✔️ 
- Descriptive PR ✔️ 
- Correct Git flow ✔️ 
- Use devise gem for authentication. ✔️ 

Good luck! 🎉 


Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**Pull request merged**